### PR TITLE
re-save blobs when saving error forms

### PR DESCRIPTION
**Pull request by snopoke**
New PR from https://github.com/dimagi/commcare-hq/pull/26655

##### SUMMARY
When an error happens after the form is saved the transaction is rolled back. This has some effects:
* form is not saved to db but `form.pk` is not cleared
* form attachments are written to the blob db but the blobdb meta is not saved
* although the blob meta isn't saved their 'saved' state is kept on the form object so when saving the error form the blobs are not re-saved

This PR addresses those by:
1. clear `model.pk` if there is a DB error
2. Separate committing of saved attachment state from the actual saving and only commit if the transactions was successful (using `transaction.on_commit`

### CODE REVIEW

**Comment by millerdev** [comment]
How is this different from what I originally reviewed in #26655? Should I do another full review?

**Comment by snopoke** [comment]
@millerdev it is identical with the addition of the last 3 commits (from [use itertools](https://github.com/dimagi/commcare-hq/pull/26693/commits/02b717b0cc67df5ee4f7f88a0ae3c1ce797dd383))

**Review by millerdev — APPROVED** [review]

**Review by millerdev — APPROVED** [review]

File: corehq/form_processor/models.py [hunk 1]
```diff
@@ -251,18 +251,27 @@ def noop_context():
...
+                self.form._attachments_list = self.saved_attachments
 
         @contextmanager
         def atomic_attachments():
             unsaved = self.attachments_list
             assert all(isinstance(a, Attachment) for a in unsaved), unsaved
             with AtomicBlobs(get_blob_db()) as blob_db:
-                yield lambda: write_attachments(blob_db)
+                writer = Writer(self, blob_db)
+                yield writer
```
[the remark this answers is not in the document]
Comment (snopoke) [reply corehq/form_processor/models.py]: :+1: a7b88724ddc

File: corehq/apps/hqcase/tests/test_object_cache.py [hunk 2]
```diff
@@ -45,18 +45,17 @@ class CaseObjectCacheTest(BaseCaseMultimediaTest):
...
     def setUp(self):
+        super(CaseObjectCacheTest, self).setUp()
         self.domain = Domain.get_or_create_with_name(TEST_DOMAIN_NAME, is_active=True)
         self.user = WebUser.create(TEST_DOMAIN_NAME, TEST_USER, TEST_PASSWORD)
         self.user.set_role(self.domain.name, 'admin')
         self.user.save()
         self.interface = FormProcessorInterface(TEST_DOMAIN_NAME)
-        delete_all_cases()
-        delete_all_xforms()
-        time.sleep(1)
```
Comment (millerdev) [inline corehq/apps/hqcase/tests/test_object_cache.py]: ⌛️ 😮

**Pull request merged**